pkg/mutation/schema/schema.go

```diff
@@ -0,0 +1,398 @@
+package schema
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// TODO we are assuming that remove will never get corrupted.
+// this should be the case, but is that wise?
+
+// Binding represent the specific GVKs that a
+// mutation's implicit schema applies to
+type Binding struct {
+	Groups   []string
+	Kinds    []string
+	Versions []string
+}
+
+// MutatorWithSchema is a mutator exposing the implied
+// schema of the target object.
+type MutatorWithSchema interface {
+	types.Mutator
+	SchemaBindings() []Binding
+}
+
+var (
+	log = logf.Log.WithName("mutation_schema")
+)
+
+type bindingIterator struct {
+	sorted bool
+}
+
+func (bi *bindingIterator) over(bindings []Binding) func() *schema.GroupVersionKind {
+	// deduplicate GVKs
+	gvksMap := map[schema.GroupVersionKind]struct{}{}
+	for _, binding := range bindings {
+		for _, group := range binding.Groups {
+			for _, version := range binding.Versions {
+				for _, kind := range binding.Kinds {
+					gvk := schema.GroupVersionKind{
+						Group:   group,
+						Version: version,
+						Kind:    kind,
+					}
+					gvksMap[gvk] = struct{}{}
+				}
+			}
+		}
+	}
+
+	gvks := []schema.GroupVersionKind{}
+	for gvk := range gvksMap {
+		gvks = append(gvks, gvk)
+	}
+
+	if bi.sorted {
+		sort.SliceStable(gvks, func(i, j int) bool { return gvks[i].String() < gvks[j].String() })
```

Comment:
Do we care about using the `Stable` flavour? gvks will be filled in random order anyway.

Reply:
We need sorting when iterating over the map in order to guarantee test coverage. Some edge cases are order-dependent, such as making sure a schema is partially applied before it fails so that we are hitting the unwinding logic.

Reply:
Can we just use `sort.Slice` ? My understanding is that the stable version leaves the order in case of equal items, but here we are popping out a map so the order is not guaranteed anyway.